File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode halfHead = head;
        ListNode temp = head;
        int n = 0, max = 0;

        while (temp != null) {
            temp = temp.next;
            n++;
        }
        // if (n == 2) return head.val + head.next.val;

        // System.out.println(n);

        for (int x = 0; x < n / 2; x++) halfHead = halfHead.next;
        // System.out.println(halfHead.val);

        // Reversing the second half of the list :
        ListNode curr = halfHead, last = halfHead;
        while(curr != null) {
            ListNode next = curr.next;
            // curr = next;
            curr.next = last;
            last = curr;
            curr = next;
        }
        // System.out.println(last.val);
        // System.out.println(last.next.val);
        for (int x = 0; x < n / 2; x++) {
            // System.out.println(head.val + " " + last.val);
            max = Math.max(head.val + last.val, max);
            head = head.next;
            last = last.next;
        }

        return max;
    }
}

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        n, temp, twin_max = 0, head, 0

        while(temp is not None) :
            temp = temp.next
            n += 1

        half_head = head
        for x in range(int(n / 2)) :
            half_head = half_head.next

        curr, last = half_head, half_head
        while curr is not None :
            consequent = curr.next
            curr.next = last
            last = curr
            curr = consequent

        for x in range(int(n / 2)) :
            twin_max = max(twin_max, head.val + last.val)
            head = head.next
            last = last.next

        return twin_max